core/src/schema.rs

```diff
@@ -0,0 +1,112 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use ::{
+    Attribute,
+    Entid,
+    Schema,
+};
+use edn::NamespacedKeyword;
+
+impl Schema {
+    pub fn get_ident(&self, x: Entid) -> Option<&NamespacedKeyword> {
+        self.entid_map.get(&x)
+    }
+
+    pub fn get_entid(&self, x: &NamespacedKeyword) -> Option<Entid> {
+        self.ident_map.get(x).map(|x| *x)
+    }
+
+    pub fn attribute_for_entid(&self, x: Entid) -> Option<&Attribute> {
+        self.schema_map.get(&x)
+    }
+
+    pub fn attribute_for_ident(&self, ident: &NamespacedKeyword) -> Option<&Attribute> {
+        self.get_entid(&ident)
+            .and_then(|x| self.attribute_for_entid(x))
+    }
+
+    /// Return `true` if the provided entid identifies an attribute in this schema.
+    pub fn is_attribute(&self, x: Entid) -> bool {
+        self.schema_map.contains_key(&x)
+    }
+
+    /// Return `true` if the provided ident identifies an attribute in this schema.
+    pub fn identifies_attribute(&self, x: &NamespacedKeyword) -> bool {
+        self.get_entid(x).map(|e| self.is_attribute(e)).unwrap_or(false)
+    }
+
+    /// Associate the given ident with given entid if neither are already recognized.
+    ///
+    /// Return `true` if an association was added, `false` otherwise.
+    pub fn associate_ident(&mut self, i: NamespacedKeyword, e: Entid) -> bool {
+        use std::collections::btree_map::Entry;
+        match (self.entid_map.entry(e), self.ident_map.entry(i.clone())) {
+            (Entry::Vacant(entid_entry), Entry::Vacant(ident_entry)) => {
```

Comment:
You can get ever so slightly better performance via:

```
if let Entry::Vacant(entid_entry) = self.entid_map.entry(e) {
    if let Entry::Vacant(…) = … {
        entid_entry.insert(i);
        ident_entry.insert(e);
        return true;
    }
}
false
```

— don't bother looking up the second if the first isn't vacant.